// syncpack/dist/lib/log-group-header.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logGroupHeader = void 0;
const chalk_1 = __importDefault(require("chalk"));
exports.logGroupHeader = {
    semverGroup(group, i) {
        log(group, 'Semver', i);
    },
    versionGroup(group, i) {
        log(group, 'Version', i);
    },
};
function log(group, type, i) {
    const customLabel = group.config.label;
    const labelWhenDefault = group.isCatchAll === true ? `Default ${type} Group` : '';
    const anonymousLabel = `${type} Group ${i + 1}`;
    const label = (customLabel || labelWhenDefault || anonymousLabel).trim();
    const hasNewLines = label.search(/[\r\n]/) !== -1;
    const header = hasNewLines ? formatMultiLine(label) : formatSingleLine(label);
    console.log(chalk_1.default.blue(header));
}
function formatSingleLine(label) {
    const leftSide = `= ${label} `;
    const dividerWidth = 80 - leftSide.length;
    const rightSide = dividerWidth > 0 ? '='.repeat(dividerWidth) : '';
    return `${leftSide}${rightSide}`;
}
function formatMultiLine(label) {
    const reindented = label.replace(/^\s+/gm, '  ');
    return `= ${reindented}`;
}
