// @adyen/api-library/lib/src/services/balancePlatform/pINFunctionalityApi.js
"use strict";
/*
 * The version of the OpenAPI document: v2
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit this class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PINFunctionalityApi = void 0;
const getJsonResponse_1 = __importDefault(require("../../helpers/getJsonResponse"));
const service_1 = __importDefault(require("../../service"));
const models_1 = require("../../typings/balancePlatform/models");
const resource_1 = __importDefault(require("../resource"));
class PINFunctionalityApi extends service_1.default {
    constructor(client) {
        super(client);
        this.API_BASEPATH = "https://balanceplatform-api-test.adyen.com/bcl/v2";
        this.baseUrl = this.createBaseUrl(this.API_BASEPATH);
    }
    /**
    * @summary Change Pin
    * @param pinChangeRequest {@link PinChangeRequest }
    * @param requestOptions {@link IRequest.Options }
    * @return {@link PinChangeResponse }
    */
    async changePin(pinChangeRequest, requestOptions) {
        const endpoint = `${this.baseUrl}/pins/change`;
        const resource = new resource_1.default(this, endpoint);
        const request = models_1.ObjectSerializer.serialize(pinChangeRequest, "PinChangeRequest");
        const response = await (0, getJsonResponse_1.default)(resource, request, { ...requestOptions, method: "POST" });
        return models_1.ObjectSerializer.deserialize(response, "PinChangeResponse");
    }
    /**
    * @summary Get RSA publicKey
    * @param requestOptions {@link IRequest.Options }
    * @param purpose {@link string } Purpose of publicKey.
    * @param format {@link string } Format of publicKey.
    * @return {@link PublicKeyResponse }
    */
    async getRsaPublickey(purpose, format, requestOptions) {
        const endpoint = `${this.baseUrl}/pins/publicKey`;
        const resource = new resource_1.default(this, endpoint);
        const hasDefinedQueryParams = purpose !== null && purpose !== void 0 ? purpose : format;
        if (hasDefinedQueryParams) {
            if (!requestOptions)
                requestOptions = {};
            if (!requestOptions.params)
                requestOptions.params = {};
            if (purpose)
                requestOptions.params["purpose"] = purpose;
            if (format)
                requestOptions.params["format"] = format;
        }
        const response = await (0, getJsonResponse_1.default)(resource, "", { ...requestOptions, method: "GET" });
        return models_1.ObjectSerializer.deserialize(response, "PublicKeyResponse");
    }
    /**
    * @summary Reveal Pin
    * @param revealPinRequest {@link RevealPinRequest }
    * @param requestOptions {@link IRequest.Options }
    * @return {@link RevealPinResponse }
    */
    async revealPin(revealPinRequest, requestOptions) {
        const endpoint = `${this.baseUrl}/pins/reveal`;
        const resource = new resource_1.default(this, endpoint);
        const request = models_1.ObjectSerializer.serialize(revealPinRequest, "RevealPinRequest");
        const response = await (0, getJsonResponse_1.default)(resource, request, { ...requestOptions, method: "POST" });
        return models_1.ObjectSerializer.deserialize(response, "RevealPinResponse");
    }
}
exports.PINFunctionalityApi = PINFunctionalityApi;
