// syncpack/dist/get-package-json-files/package-json-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageJsonFile = void 0;
const Function_1 = require("@effect/data/Function");
const path_1 = require("path");
const map_1 = require("tightrope/result/map");
const constants_1 = require("../constants");
const create_1 = require("../instance/create");
const log_verbose_1 = require("../lib/log-verbose");
class PackageJsonFile {
    constructor(jsonFile, config) {
        this.config = config;
        this.contents = jsonFile.contents;
        this.filePath = jsonFile.filePath;
        this.dirPath = (0, path_1.dirname)(jsonFile.filePath);
        this.json = jsonFile.json;
        this.shortPath = (0, path_1.relative)(constants_1.CWD, jsonFile.filePath);
    }
    getInstances(enabledTypes) {
        const instances = [];
        enabledTypes.forEach((strategy) => {
            (0, Function_1.pipe)(strategy.read(this), (0, map_1.map)((entries) => entries.forEach(([name, specifier]) => {
                (0, log_verbose_1.logVerbose)(`add ${name}@${specifier} to ${strategy.name}:${strategy._tag} ${this.shortPath}`);
                instances.push((0, create_1.createInstance)(strategy, name, this, specifier));
            })));
        });
        return instances;
    }
}
exports.PackageJsonFile = PackageJsonFile;
