// syncpack/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONFIG = exports.RANGE = exports.ICON = exports.DEFAULT_SOURCES = exports.CWD = void 0;
/** Single source of truth, intended to aid testing or to override */
exports.CWD = process.env.MOCK_CWD || process.cwd();
/** Where to search for packages if none are provided by the user */
exports.DEFAULT_SOURCES = ['package.json', 'packages/*/package.json'];
/** Single source of truth for icons used in output */
exports.ICON = {
    cross: '✘',
    debug: '?',
    panic: '!',
    rightArrow: '→',
    skip: '-',
    tick: '✓',
};
exports.RANGE = {
    ANY: '*',
    EXACT: '',
    GT: '>',
    GTE: '>=',
    LOOSE: '.x',
    LT: '<',
    LTE: '<=',
    MINOR: '^',
    PATCH: '~',
    WORKSPACE: 'workspace:',
};
exports.DEFAULT_CONFIG = {
    dependencyTypes: [
        'dev',
        'overrides',
        'peer',
        'pnpmOverrides',
        'prod',
        'resolutions',
        'workspace',
    ],
    filter: '.',
    indent: '  ',
    semverGroups: [],
    semverRange: '',
    sortAz: [
        'contributors',
        'dependencies',
        'devDependencies',
        'keywords',
        'peerDependencies',
        'resolutions',
        'scripts',
    ],
    sortFirst: ['name', 'description', 'version', 'author'],
    source: exports.DEFAULT_SOURCES,
    versionGroups: [],
};
