// syncpack/dist/get-version-groups/same-range.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SameRangeVersionGroup = void 0;
const Data = __importStar(require("@effect/data/Data"));
const Effect = __importStar(require("@effect/io/Effect"));
const intersects_1 = __importDefault(require("semver/ranges/intersects"));
const _1 = require(".");
const group_by_1 = require("./lib/group-by");
class SameRangeVersionGroup extends Data.TaggedClass('SameRange') {
    constructor(config) {
        super({
            config,
            instances: [],
        });
    }
    canAdd(_) {
        return true;
    }
    inspect() {
        const instancesByName = (0, group_by_1.groupBy)('name', this.instances);
        return Object.entries(instancesByName).map(([name, instances]) => {
            if (hasMismatch(instances)) {
                return Effect.fail(new _1.VersionGroupReport.SameRangeMismatch({
                    name,
                    instances,
                    isValid: false,
                }));
            }
            else {
                return Effect.succeed(new _1.VersionGroupReport.Valid({
                    name,
                    instances,
                    isValid: true,
                }));
            }
        });
    }
}
exports.SameRangeVersionGroup = SameRangeVersionGroup;
/** Every range must fall within every other range */
function hasMismatch(instances) {
    const loose = true;
    return instances.some((a) => instances.some((b) => !(0, intersects_1.default)(aliasWorkspaceRange(a.specifier), aliasWorkspaceRange(b.specifier), loose)));
}
function aliasWorkspaceRange(version) {
    return version.startsWith('workspace:*') ? '*' : version;
}
