// syncpack/dist/get-version-groups/snapped-to.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnappedToVersionGroup = void 0;
const Data = __importStar(require("@effect/data/Data"));
const Effect = __importStar(require("@effect/io/Effect"));
const _1 = require(".");
const group_by_1 = require("./lib/group-by");
class SnappedToVersionGroup extends Data.TaggedClass('SnappedTo') {
    constructor(config) {
        super({
            config,
            instances: [],
        });
    }
    canAdd(_) {
        return true;
    }
    inspect() {
        const instancesByName = (0, group_by_1.groupBy)('name', this.instances);
        return Object.entries(instancesByName).map(([name, instances]) => {
            const snapTo = this.config.snapTo;
            const expectedVersion = getExpectedVersion(snapTo, instances);
            if (hasMismatch(expectedVersion, instances)) {
                return Effect.fail(new _1.VersionGroupReport.SnappedToMismatch({
                    name,
                    instances,
                    isValid: false,
                    expectedVersion,
                    snapTo,
                }));
            }
            else {
                return Effect.succeed(new _1.VersionGroupReport.Valid({
                    name,
                    instances,
                    isValid: true,
                }));
            }
        });
    }
}
exports.SnappedToVersionGroup = SnappedToVersionGroup;
function getExpectedVersion(snapTo, instances) {
    var _a;
    const expectedVersion = (_a = instances
        .filter((i) => snapTo.includes(i.pkgName))
        .find((i) => i.specifier)) === null || _a === void 0 ? void 0 : _a.specifier;
    if (expectedVersion)
        return expectedVersion;
    // @FIXME: create tagged error for this
    throw new Error('versionGroup.snapTo does not match any package versions');
}
function hasMismatch(expectedVersion, instances) {
    return instances.some((instance) => instance.specifier !== expectedVersion);
}
